File: src/128. Longest Consecutive Sequence.cpp
/*Problem link: https://leetcode.com/problems/longest-consecutive-sequence/
 *Problem Difficulty: Hard
 *Time Complexity: O(N)
 *Space Complexity: O(N)
*/ 

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        int range = 0, ans = 0;
        unordered_map<int, bool> m;
        for (auto &u : nums) {
            m[u] = false;
        }
        for (auto &u : m) {
            if (u.second == true) continue;
            u.second = true;
            range = 1; 
            int n = u.first;
            while(m.find(n - 1) != m.end()) {
                ++range;
                m[n - 1] = true;
                --n;
            }
            n = u.first;
            while(m.find(n + 1) != m.end()) {
                ++range;
                m[n + 1] = true;
                ++n;
            }
            ans = max(ans, range);
        }
        return ans;
    }
};

File: 128. Longest Consecutive Sequence.java
/* https://leetcode.com/problems/longest-consecutive-sequence/ */

class Solution {
    public int longestConsecutive(int[] nums) {
        int ans = 0;
        if (nums.length == 0)
            return ans;
        HashMap<Integer, Boolean> consequtive = new HashMap<Integer, Boolean>();
        for (int i = 0; i < nums.length; i++) {
            consequtive.put(nums[i], false);
        }
        for (int num : nums) {
            if (consequtive.get(num) == true)
                continue;
            consequtive.put(num, true);
            int current = num, current_length = 1;
            while (consequtive.get(current - 1) != null) {
                current_length++;
                current--;
            }
            current = num;
            while (consequtive.get(current + 1) != null) {
                current_length++;
                current++;
            }
            ans = Math.max(ans, current_length);
        }
        return ans;
    }
}
